#sprint_9

Anton8P opened the pull request:
feat: реализовать HTTP API для трекера задач
- Добавлен HTTP сервер (HttpTaskServer)
- Созданы обработчики для эндпоинтов: /tasks, /subtasks, /epics, /history, /prioritized
- Реализован BaseHttpHandler с обработкой ошибок
- Настроена JSON сериализация через Gson с адаптером для LocalDateTime
- Добавлены DTO класс и конвертер для обмена данными
- Добавлена обработка исключений (NotFoundException)
- Добавлены тесты для всех эндпоинтов HTTP API
- Обновлены методы менеджера для работы с исключениями вместо возврата null

Commit:
feat: реализовать HTTP API для трекера задач

- Добавлен HTTP сервер (HttpTaskServer)
- Созданы обработчики для эндпоинтов: /tasks, /subtasks, /epics, /history, /prioritized
- Реализован BaseHttpHandler с обработкой ошибок
- Настроена JSON сериализация через Gson с адаптером для LocalDateTime
- Добавлены DTO класс и конвертер для обмена данными
- Добавлена обработка исключений (NotFoundException)
- Добавлены тесты для всех эндпоинтов HTTP API
- Обновлены методы менеджера для работы с исключениями вместо возврата null

Ivanov-s-tmn commented on .idea/codeStyles/codeStyleConfig.xml:
Директория .idea генерируется локально и не нужна в репозитории, ее стоит добавить в исключения gitignore

Ivanov-s-tmn commented on src/server/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,185 @@
+
+
+
+
```
Куча лишних пустых строк)

Ivanov-s-tmn commented on src/server/handlers/BaseHttpHandler.java:
`BaseHttpHandler` можно сделать абстрактным и имплементировать интерфейс `HttpHandler` + вынести сюда общее поле менеджера

Ivanov-s-tmn commented on src/server/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,185 @@
+            if (path.equals("/epics")) {
+                String body = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
+
+                try {
+                    TaskDto taskDto = GSON.fromJson(body, TaskDto.class);
+
+                    if (taskDto.getTitle() == null || taskDto.getTitle().isBlank()) {
+                        sendIncorrectRequest(exchange);
+                        return;
+                    }
+
+                    if (taskDto.getType() == null) {
+                        taskDto.setType(TaskType.EPIC);
+                    }
+
+                    if (taskDto.getType() != TaskType.EPIC) {
+                        sendIncorrectRequest(exchange);
+                        return;
+                    }
+
+                    Task task = TaskDtoConverter.toEntity(taskDto);
+
+                    Epic epic = (Epic) task;
+
+                    int id = taskManager.createEpic(epic);
+                    taskDto.setId(id);
+                    sendText(exchange, GSON.toJson(taskDto), 201);
+
+                } catch (JsonSyntaxException e) {
+                    sendIncorrectRequest(exchange);
+                } catch (NotFoundException e) {
+                    sendNotFound(exchange);
+                }
+            } else {
+                sendNotFound(exchange);
+            }
+        } catch (Exception e) {
+            sendInternalError(exchange);
+        }
+    }
```
Помимо создание должно быть еще обновление

Ivanov-s-tmn commented on test/server/HttpTaskServerEpicsTest.java:
```diff
@@ -0,0 +1,61 @@
+                    "title": "Test Epic",
+                    "type": "EPIC"
+                }
+                """;
```
Стоит преобразовывать через gson методом `toJson()`, писать строку не так удобно и стабильно

Ivanov-s-tmn commented on test/server/HttpTaskServerEpicsTest.java:
Нужно покрыть тестами все эндпоинты

Ivanov-s-tmn commented on test/server/HttpTaskServerSubtasksTest.java:
```diff
@@ -0,0 +1,109 @@
+                .uri(URI.create(urlBase + "/subtasks"))
+                .POST(HttpRequest.BodyPublishers.ofString(subtaskJson))
+                .header("Content-Type", "application/json")
+                .build();
```
Это тесты для эндпоинта получения сабтаски по айди, соответственно в тесте стоит делать только этот GET запрос на /substasks, а добавлять тестовые объекты через сам менеджер.

1 тест - 1 запрос

Commit:
Правки после ревью:
- изменены тесты для HTTP сервера
- все эндпоинты покрыты тестами
- базовый класс для обработчиков объявлен абстрактным
- директория .idea добавлена в gitignore

Ivanov-s-tmn commented on src/server/handlers/EpicsHandler.java:
```diff
@@ -66,6 +62,7 @@ private void handleGet(HttpExchange exchange, String path) throws IOException {
                     if (isNumber(strIdWithSubtasks)) {
                         int idWithSubtasks = Integer.parseInt(strIdWithSubtasks);
                         try {
+                            taskManager.getEpicById(idWithSubtasks);
```
Этот лишний вызов `taskManager.getEpicById(idWithSubtasks)` тут не нужен, у нас же в таск менеджере заложена логика добавления в историю, а это эндпоинт получения подзадач эпика, а не просмотр самого эпика. Если он здесь для отлова `NotFoundException`, то он должен быть по хорошему в логике `getAllSubtasksByEpicId`

Anton8P replied:
исправил. добавил выброс исключения в метод getAllSubtasksByEpicId(int epicId) в менеджере.

Commit:
в метод getAllSubtasksByEpicId(int epicId) в менеджере добавлен выброс исключения, если эпика нет в мапе
